components/server/src/authorization/spicedb-authorizer.ts

```diff
@@ -196,3 +228,111 @@ export class SpiceDBAuthorizerImpl implements SpiceDBAuthorizer {
         }) as any as grpc.Metadata;
     }
 }
+
+type ZedTokenCacheKV = [objectRef: v1.ObjectReference | undefined, token: string | undefined];
+interface ZedTokenCache {
+    get(objectRef: v1.ObjectReference): Promise<string | undefined>;
+    set(...kvs: ZedTokenCacheKV[]): Promise<void>;
+    consistency(resourceRef: v1.ObjectReference | undefined): Promise<v1.Consistency>;
+}
+
+type ContextWithZedToken = RequestContext & { zedToken?: StoredZedToken };
+function getContext(): ContextWithZedToken {
+    return getRequestContext() as ContextWithZedToken;
+}
+
+/**
+ * This is a simple implementation of the ZedTokenCache that uses the local context to store single ZedToken per API request, which is stored in AsyncLocalStorage.
+ * To make this work we make the "assumption" that ZedTokens string (meant to be opaque) represent a timestamp which we can order. This is at least true for the MySQL datastore we are using.
+ */
+export class RequestLocalZedTokenCache implements ZedTokenCache {
+    constructor() {}
+
+    async get(objectRef: v1.ObjectReference | undefined): Promise<string | undefined> {
+        return getContext().zedToken?.token;
+    }
+
+    async set(...kvs: ZedTokenCacheKV[]) {
+        function clearZedTokenOnContext() {
+            getContext().zedToken = undefined;
+        }
+
+        const mustClearCache = kvs.some(([k, v]) => !!k && !v); // did we write a relationship without getting a writtenAt token?
+        if (mustClearCache) {
+            clearZedTokenOnContext();
+            return;
+        }
+
+        try {
+            const allTokens = [
+                ...kvs.map(([_, v]) => (!!v ? StoredZedToken.fromToken(v) : undefined)),
+                getContext().zedToken,
+            ].filter((v) => !!v) as StoredZedToken[];
+            const freshest = this.freshest(...allTokens);
+            if (freshest) {
+                getContext().zedToken = freshest;
+            }
+        } catch (err) {
+            log.warn("[spicedb] Failed to set ZedToken on context", err);
+            clearZedTokenOnContext();
+        }
+    }
+
+    async consistency(resourceRef: v1.ObjectReference | undefined): Promise<v1.Consistency> {
```

Comment:
nit: I would pass in the zedtoken and move this method to the authorizer. it seems unrelated to the caching aspect.

Reply:
For this cache implementation: Yes. For others definitely not. Again, happy to remove/fold in if we don't need that stuff. :crossed_fingers: